### DISCUSSION TITLE
Getting the location (coordinates) of the aircraft

### DISCUSSION THREAD

**Discussion by LuFlo** [JSBSim-Team/jsbsim#638] [Q&A]
The programmer documentation lacks some really basic information. I mean, come on guys, this should be documented :( I found this: http://jsbsim.sourceforge.net/JSBSim/classJSBSim_1_1FGLocation.html
But nowhere is said WHERE to get this object. From FGAircraft? No. Directly from the executor? No. Also I am using the python wrapper and somehow there are a lot of public members missing in FGAircraft, maybe the position is there?

**Comment by seanmcleod** [comment]
> I mean, come on guys, this should be documented :( 

Well pull requests are welcome :wink:

In terms of getting location data take a look at:

https://github.com/JSBSim-Team/jsbsim/blob/59389f66c790d0a8ae2323dbedbbf9b278e32e2c/src/models/FGPropagate.cpp#L845-L868

So you could get hold of `FGPropagate` or query based on the property, e.g. in python.

```python
fdm = jsbsim.FGFDMExec('..\\') # The path supplied to FGFDMExec is the location of the folders "aircraft", "engines" and "systems"
fdm.load_model('737') 

lon = fdm['position/long-gc-rad']
```

**Reply by LuFlo** [reply]
Thanks for the fast response! Yes, I definitely will consider doing some pull request as soon as I get a hang of JSBsim :D
Unfortunately the python wrapper seems not to include all definitions for FGPropagate, I tried for example this:

```python
print(fdm.get_propagate().get_location())
```

But it only produces an error: `AttributeError: 'jsbsim.FGPropagate' object has no attribute 'get_location'`

I managed to get the position via the dictionary access as you showed in the example.

One more question: Where do I find definitions / documentation for the properties? For example, what is "ecef-x-ft"? Is it the location in xyz (with origin in the middle of geod) in feet? Then whats the difference between "InertialPosition" and "Location"?

**Comment by seanmcleod** [comment]
> Where do I find definitions / documentation for the properties?

Often there are useful comments in the source, e.g.

https://github.com/JSBSim-Team/jsbsim/blob/59389f66c790d0a8ae2323dbedbbf9b278e32e2c/src/models/FGPropagate.h#L101-L104

https://github.com/JSBSim-Team/jsbsim/blob/59389f66c790d0a8ae2323dbedbbf9b278e32e2c/src/models/FGPropagate.h#L316-L318

There is also some specific documentation with regards to frames of reference:

https://jsbsim-team.github.io/jsbsim-reference-manual/mypages/user-manual-frames-of-reference/

**Comment by agodemar** [comment]
@LuFlo 

> The programmer documentation lacks some really basic information. I mean, come on guys, this should be documented :( I found this: http://jsbsim.sourceforge.net/JSBSim/classJSBSim_1_1FGLocation.html
But nowhere is said WHERE to get this object. From FGAircraft? No. Directly from the executor? No. Also I am using the python wrapper and somehow there are a lot of public members missing in FGAircraft, maybe the position is there?

The JSBSim software library is the result of years of volunteer work. And you have all you need at your finger tips. For free.
It would be great if smart people having a look at this amazing piece of software could help us out to improve the APIs and the documentation.

**Reply by LuFlo** [reply]
I don't doubt that JSBsim is amazing software and that a lot of developer put years and years of their time into this project. I never stated the opposite. I also know very well what open source software is, I work with free and open source software every day, in my profession as well as in my free time. Yes maybe I should have formulated it differently since in text form it's easy to get the wrong impression.

I still do think that the criticism is valid though. And I don't even say that the project doesn't have documentation, quite contrary. The PDF document has 160 pages or so. But the core of a flight simulation library is to calculate position and attitude of an aircraft given some input variables. This key part (getting the core information) wasn't very obvious for me, as a person who never used JSBsim before. I would be very happy to help documenting this part, I've seen that you guys have a separate repository for the user manual. I already put some example files together which I used for my recent bug reports / discussions, these can be a basis for some useful documentation for common pitfalls. What do you think?

**Reply by agodemar** [reply]
@LuFlo that would be a great way to get involved. Thanks, also for the criticism, as in this case it brings the attention on a pitfall that can be easily treated.
There are many little things in open source projects that are as important as code. A clear documentation, of course, but also the deployment workflow as well as reproducible examples of use.
In this last area, I think Jupyter notebooks could be a handy tool.

**Comment by seanmcleod** [comment]
> I've seen that you guys have a separate repository for the user manual.

Yep, this was to serve as a replacement/update of the original PDF document, but it's been stalled for a couple of years.

**Reply by LuFlo** [reply]
I think I just found out why 😅 apparently it is surprisingly difficult to get Jekyll running on modern distributions, since it relies on ruby 2.7, see here for example: https://github.com/rbenv/ruby-build/discussions/1940#discussioncomment-2585908

I think, it's still manageable in VMs which run older distributions, buts kinda painful for local development. I will take a closer look the next days.

**Reply by seanmcleod** [reply]
Nah, it stalled long before installation issues became an issue :wink:

**Reply by LuFlo** [reply]
Well, that's unfortunate 😅 still, solving installation issues is worth trying :)

**Reply by agodemar** [reply]
@seanmcleod @bcoconni a better alternative to the technology that I chose more than 3 years ago for the online documentation is [Docusaurus](https://www.docusaurus.io/). It is supported by GitHub as well and relies on Node.js (no Jekyll).

**Reply by LuFlo** [reply]
@agodemar I opened an [issue](https://github.com/JSBSim-Team/jsbsim-reference-manual/issues/22) on the reference manual repository for this. I also figured out a [workaround](https://github.com/JSBSim-Team/jsbsim-reference-manual/issues/21) for local development, although it doesn't have live reloading. I think we can move the discussion about this topic there.
